### PR TITLE
[Web][Api][SIMS #915] ministry user views pending applications

### PR DESCRIPTION
**Pull request by henryoforeh-aot**
- Created Student Application Exception summary page.
- Implemented menu item to navigate to Student Applications Extensions Summary page.
- Created new service to get all Student Application Exceptions.
- Implemented navigation to view Student Applications Assessments screen.

image

### CODE REVIEW

File: sources/packages/api/src/route-controllers/application-exception/application-exception.aest.controller.ts [hunk 1]
```diff
@@ -119,4 +125,32 @@ export class ApplicationExceptionAESTController extends BaseController {
...
+    @Query() pagination: ApplicationStatusPaginationOptionsAPIInDTO,
+  ): Promise<PaginatedResultsAPIOutDTO<ApplicationExceptionSummaryAPIOutDTO>> {
+    const applicationExceptions =
+      await this.applicationExceptionService.getApplicationExceptions(
+        pagination,
+      );
+
+    return {
+      results: applicationExceptions.results.map((eachApplication) => {
+        return {
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/application-exception/application-exception.aest.controller.ts:145]: This code can be simplified as below:
```typescript
return {
      results: applicationExceptions.results.map((eachApplication) => ({
          applicationId: eachApplication.application.id,
          studentId: eachApplication.application.student.id,
          applicationNumber: eachApplication.application.applicationNumber,
          submittedDate: eachApplication.createdAt,
          fullName: getUserFullName(eachApplication.application.student.user),
        });
      ),
      count: applicationExceptions.count,
    };
```

File: sources/packages/api/src/route-controllers/application-exception/application-exception.aest.controller.ts [hunk 2]
```diff
@@ -119,4 +125,32 @@ export class ApplicationExceptionAESTController extends BaseController {
...
+   * Gets all student application exceptions.
+   * @param pagination options to execute the pagination.
+   * @returns list of student application exceptions.
+   */
+  @Get()
+  async getExceptions(
+    @Query() pagination: ApplicationStatusPaginationOptionsAPIInDTO,
+  ): Promise<PaginatedResultsAPIOutDTO<ApplicationExceptionSummaryAPIOutDTO>> {
+    const applicationExceptions =
+      await this.applicationExceptionService.getApplicationExceptions(
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/application-exception/application-exception.aest.controller.ts:139]: Maybe the ticket is not clear or I am under the wrong impression, this query is supposed to display only "Pending" exceptions, right?

Comment (henryoforeh-aot) [reply sources/packages/api/src/route-controllers/application-exception/application-exception.aest.controller.ts:139]: Correct, missed it. I will update.

File: sources/packages/web/src/views/aest/institution/StudentApplicationExceptions.vue [hunk 3]
```diff
@@ -0,0 +1,17 @@
+<template>
```
[the remark this answers is not in the document]
Comment (henryoforeh-aot) [reply sources/packages/web/src/views/aest/institution/StudentApplicationExceptions.vue:1]: Nice catch, moving it to views\aest\student.

File: sources/packages/api/src/route-controllers/application-exception/application-exception.aest.controller.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (henryoforeh-aot) [reply sources/packages/api/src/route-controllers/application-exception/application-exception.aest.controller.ts]: I have implemented as adviced.

File: sources/packages/api/src/route-controllers/application-exception/application-exception.aest.controller.ts [hunk 4]
```diff
@@ -119,4 +125,30 @@ export class ApplicationExceptionAESTController extends BaseController {
...
+  async getPendingApplicationExceptions(
+    @Query() pagination: ApplicationExceptionPaginationOptionsAPIInDTO,
+  ): Promise<PaginatedResultsAPIOutDTO<ApplicationExceptionSummaryAPIOutDTO>> {
+    const applicationExceptions =
+      await this.applicationExceptionService.getPendingApplicationExceptions(
+        pagination,
+      );
+
+    return {
+      results: applicationExceptions.results.map((eachApplication) => ({
```
Comment (andrewsignori-aot) [inline sources/packages/api/src/route-controllers/application-exception/application-exception.aest.controller.ts:144]: Just double-checking, case there are no records the value from `applicationExceptions.results` will be `null` or an empty array?

File: sources/packages/web/src/constants/routes/RouteConstants.ts [hunk 5]
```diff
@@ -104,6 +104,7 @@ export const AESTRoutesConst = {
   NOTICE_OF_ASSESSMENT_VIEW: Symbol(),
   REPORTS: Symbol(),
   APPLICATION_EXCEPTIONS_APPROVAL: Symbol(),
+  EXCEPTIONS: Symbol(),
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/constants/routes/RouteConstants.ts:107]: Can we follow the same one from the above one and call it `APPLICATION_EXCEPTIONS_PENDING` or similar?

File: sources/packages/web/src/services/http/ApplicationExceptionApi.ts [hunk 6]
```diff
@@ -38,4 +42,37 @@ export class ApplicationExceptionApi extends HttpBaseClient {
...
+   * Gets all pending student application exceptions.
+   * @param paginationOptions options to execute the pagination.
+   * @returns list of student application exceptions.
+   */
+  async getPendingExceptions(
+    paginationOptions: PaginationOptions,
+  ): Promise<PaginatedResultsAPIOutDTO<ApplicationExceptionSummaryAPIOutDTO>> {
+    let url = `application-exception`;
+    // Adding pagination params. There is always a default page and pageLimit for paginated APIs.
+    url = addPaginationOptions(
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/services/http/ApplicationExceptionApi.ts:56]: **Not part of this PR** but the code from line 56 till 73 should be moved to a helper near `addPaginationOptions` and `addSortOptions` and just receive the URL + paginationOptions.

File: sources/packages/web/src/views/aest/student/StudentApplicationExceptions.vue [hunk 7]
```diff
@@ -0,0 +1,173 @@
+<template>
+  <full-page-container>
+    <template #header>
+      <header-navigator title="Students" subTitle="Exceptions" />
+    </template>
+
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/aest/student/StudentApplicationExceptions.vue:6]: IMO these blank limes in the middle of the HTML are not bringing better readability hence they can be removed.

File: sources/packages/web/src/views/aest/student/StudentApplicationExceptions.vue [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+  DEFAULT_PAGE_NUMBER,
+  PageAndSortEvent,
+  PaginatedResults,
+} from "@/types";
+
+import { useFormatters } from "@/composables";
+import { AESTRoutesConst } from "@/constants/routes/RouteConstants";
+import { ApplicationExceptionSummaryAPIOutDTO } from "@/services/http/dto/ApplicationException.dto";
+
+const DEFAULT_SORT_FIELD = "applicationNumber";
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/aest/student/StudentApplicationExceptions.vue:97]: IMO the default should be the submittedDate ordered ASC.

File: sources/packages/web/src/views/aest/student/StudentApplicationExceptions.vue [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+        <v-text-field
+          class="v-text-field-search-width"
+          density="compact"
+          label="Search name or application #"
+          variant="outlined"
+          v-model="searchCriteria"
+          @keyup.enter="searchExceptions"
+        >
+          <template v-slot:prependInner>
+            <font-awesome-icon :icon="['fas', 'search']" class="m" />
```
Comment (ann-aot) [inline sources/packages/web/src/views/aest/student/StudentApplicationExceptions.vue:23]: does this class `m` have any effect?

File: sources/packages/web/src/views/aest/student/StudentApplicationExceptions.vue [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+} from "@/types";
+
+import { useFormatters } from "@/composables";
+import { AESTRoutesConst } from "@/constants/routes/RouteConstants";
+import { ApplicationExceptionSummaryAPIOutDTO } from "@/services/http/dto/ApplicationException.dto";
+
+const DEFAULT_SORT_FIELD = "applicationNumber";
+
+export default {
+  setup(_props: any) {
```
Comment (andrewsignori-aot) [inline sources/packages/web/src/views/aest/student/StudentApplicationExceptions.vue:100]: `_props` can be removed.

**Review by andrewsignori-aot — COMMENTED** [review]
Thanks for doing the changes, please take a look at the comments.

File: sources/packages/web/src/views/aest/student/StudentApplicationExceptions.vue (hunk 10 above)
Comment (ann-aot) [inline sources/packages/web/src/views/aest/student/StudentApplicationExceptions.vue:100]: if the `props` is not used. then no need to add it

File: sources/packages/web/src/views/aest/student/StudentApplicationExceptions.vue (hunk 9 above)
Comment (henryoforeh-aot) [reply sources/packages/web/src/views/aest/student/StudentApplicationExceptions.vue:23]: I will remove it.

File: sources/packages/api/src/route-controllers/application-exception/application-exception.aest.controller.ts (hunk 4 above)
Comment (henryoforeh-aot) [reply sources/packages/api/src/route-controllers/application-exception/application-exception.aest.controller.ts:144]: An empty array will be returned. 


image

**Review by ann-aot — APPROVED** [review]
Good work @henryoforeh-aot . Thanks for doing the changes 👍

**Review by guru-aot — APPROVED** [review]
LGTM, NIce work @henryoforeh-aot
